quinn/src/notify.rs

```diff
@@ -0,0 +1,89 @@
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Waker};
+
+use slab::Slab;
+
+/// Broadcasts an event to any number of waiters
+#[derive(Clone)]
+pub struct NotifyOwned {
+    shared: Arc<Mutex<Shared>>,
+}
+
+struct Shared {
+    version: u64,
+    wakers: Slab<Waker>,
+}
+
+impl NotifyOwned {
+    pub fn new() -> Self {
+        Self {
+            shared: Arc::new(Mutex::new(Shared {
+                version: 0,
+                wakers: Slab::new(),
+            })),
+        }
+    }
+
+    pub fn notify_all(&self) {
+        let mut shared = self.shared.lock().unwrap();
+        shared.version = shared.version.wrapping_add(1);
+        let n = shared.wakers.len();
+        for waker in shared.wakers.drain() {
+            waker.wake();
+        }
+        // Limit unused slots, which cost iteration time as well as memory
+        if n * 4 < shared.wakers.capacity() {
+            shared.wakers = Slab::with_capacity(n);
+        }
+    }
+
+    pub fn wait(&self) -> Waiter {
+        Waiter {
+            shared: self.shared.clone(),
+            state: None,
+        }
+    }
+}
+
+pub struct Waiter {
+    shared: Arc<Mutex<Shared>>,
+    state: Option<State>,
+}
+
+impl Waiter {
+    /// Register to be woken by the next `notify_all`. Must be called before readiness could occur,
```

Comment:
Short single no-period first line of docs?

Reply:
Mooted by adoption of `tokio::sync::Notify`.